hardware/plugins/Plugins.cpp

```diff
@@ -657,19 +658,64 @@ namespace Plugins {
 		if (m_bIsStarted) StopHardware();
 
 		//	Add start command to message queue
+		m_bIsStarting = true;
 		MessagePlugin(new InitializeMessage(this));
 
 		_log.Log(LOG_STATUS, "(%s) Started.", Name.c_str());
 
 		return true;
 	}
 
+	void CPlugin::ClearMessageQueue()
+	{
+		// Copy the event queue to a temporary one, then copy back the events for other plugins
+		boost::lock_guard<boost::mutex> l(PluginMutex);
+		std::queue<CPluginMessageBase*>	TempMessageQueue(PluginMessageQueue);
+		while (!PluginMessageQueue.empty())
+			PluginMessageQueue.pop();
+
+		while (!TempMessageQueue.empty())
+		{
+			CPluginMessageBase* FrontMessage = TempMessageQueue.front();
+			TempMessageQueue.pop();
+			if (FrontMessage->m_pPlugin == this)
+			{
+				// log events that will not be processed
+				CCallbackBase* pCallback = dynamic_cast<CCallbackBase*>(FrontMessage);
+				if (pCallback)
+					_log.Log(LOG_ERROR, "(%s) Callback event '%s' (Python call '%s') discarded.", Name.c_str(), FrontMessage->Name(), pCallback->PythonName());
+				else
+					_log.Log(LOG_ERROR, "(%s) Non-callback event '%s' discarded.", Name.c_str(), FrontMessage->Name());
+			}
+			else
+			{
+				// Message is for a different plugin so requeue it
+				_log.Log(LOG_NORM, "(%s) requeuing '%s' message for '%s'", Name.c_str(), FrontMessage->Name(), FrontMessage->Plugin()->Name.c_str());
+				PluginMessageQueue.push(FrontMessage);
+			}
+		}
+	}
+
 	bool CPlugin::StopHardware()
 	{
 		try
 		{
 			_log.Log(LOG_STATUS, "(%s) Stop directive received.", Name.c_str());
 
+			// loop on plugin to finish startup
```

Comment:
Can't plugins be started in blocking mode ? Then we know when 'm_bIsStarted' is true, it was started, else not... we do not need a m_bIsStarting then
Then you can use the

boost::lock_guard<boost::mutex> l(PluginMutex);

in the stophardware ?

Also in this routine 'm_Notifier' is deleted, i would also set this then to NULL after deletion

Reply:
I agree adding the m_bIsStarting complicates the logic a bit, and increases risk of new issues..
Unfortunately, I don't think starting the plugins in blocking mode is a good idea, it would mean a plugin such as the one I used for testing, https://github.com/ffes/domoticz-buienradar, would block the domoticz main thread for a very long time. That plugin does some http access at startup, and in my own testing this needs up to 30 s.

Agreed on setting to NULL after deletion.